src/relay/backend/te_compiler.cc

```diff
@@ -383,10 +384,9 @@ class LowerTensorExprMutator : public ExprMutator {
    *  \brief Returns the primitive function associated with \p expr, or
    *  nullptr if none.
    */
-  Function ResolveToPrimitive(Expr expr) {
+  BaseFunc ResolveToPrimitive(Expr expr) {
     if (const GlobalVarNode* gvn = expr.as<GlobalVarNode>()) {
-      BaseFunc base_func = module_->Lookup(GetRef<GlobalVar>(gvn));
-      return ResolveToPrimitive(base_func);
+      return module_->Lookup(GetRef<GlobalVar>(gvn));
```

Comment:
The caller is relying on this to filter out non-primitives, I think we still need this since when used by the interpreter there may still be calls to non-prims in the AST.

Reply:
The two callers of this should skip over this by virtue of the:
```
if (prim_func->IsInstance<tir::PrimFuncNode>()) {
  return GetRef<Let>(let);
}
```
I added to both of them, so it's either undefined in which case it drops into not defined or it's a `PrimFunc` which just skips the mutation.

Though there are some failing tests which seem related  - I'll dig into that tomorrow.

Reply:
Ah, sorry 'primitive' is overloaded. What I mean is the interpreter may encounter Relay functions which are not marked with the kPrimitive attribute, and those should not be lowered. That filtering used to be done by the recursive call you removed.

Reply:
Argh, I saw that and didn't clock it. Thanks for the heads up, easy fix that could've taken me ages :smile_cat: